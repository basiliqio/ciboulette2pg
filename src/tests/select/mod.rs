use super::*;

mod query_params;
mod select_success;

async fn test_select<'a>(
    transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
    query_end: &str,
    _name: &str,
    _data: &BTreeMap<String, Vec<Uuid>>,
) -> Vec<sqlx::postgres::PgRow> {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store(&ciboulette_store);
    let parsed_url = Url::parse(format!("http://localhost{}", query_end).as_str()).unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Read;

    let req_builder = CibouletteInboundRequestBuilder::new(INTENTION, &parsed_url, &None);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteReadRequest::try_from(request).unwrap();
    let builder = Ciboulette2PostgresBuilder::gen_select(
        &ciboulette_store,
        &table_store,
        &ciboulette_request,
    )
    .unwrap();
    let (query, args) = builder.build().unwrap();
    println!(
        "{}",
        sqlformat::format(
            query.as_str(),
            &sqlformat::QueryParams::None,
            sqlformat::FormatOptions::default()
        )
    );
    let raw_rows: Vec<sqlx::postgres::PgRow> = sqlx::query_with(&query, args)
        .fetch_all(&mut *transaction)
        .await
        .unwrap();
    raw_rows
}
