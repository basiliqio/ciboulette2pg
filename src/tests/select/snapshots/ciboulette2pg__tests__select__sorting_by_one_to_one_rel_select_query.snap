---
source: src/tests/select/mod.rs
expression: "sqlformat::format(query.as_str(), &sqlformat::QueryParams::None,\n                  sqlformat::FormatOptions::default())"

---
WITH "cte_peoples_data_0" AS (
  SELECT
    "public"."peoples"."id" :: TEXT AS "id",
    "public"."peoples"."id" AS "main_id",
    NULL :: TEXT AS "related_id",
    NULL :: TEXT AS "related_type",
    $1 :: TEXT AS "type",
    JSON_BUILD_OBJECT(
      $2,
      "public"."peoples"."age",
      $3,
      "public"."peoples"."first-name",
      $4,
      "public"."peoples"."gender",
      $5,
      "public"."peoples"."last-name",
      $6,
      "public"."peoples"."twitter"
    ) AS "data",
    "public"."peoples"."favorite_color" AS "rel_favorite_color",
    "public"."peoples"."first-name" AS "sort_first-name"
  FROM
    "public"."peoples"
),
"cte_rel_favorite_color_data_1" AS (
  SELECT
    "public"."favorite_color"."id" :: TEXT AS "id",
    "public"."favorite_color"."id" AS "main_id",
    "cte_peoples_data_0"."id" :: TEXT AS "related_id",
    $7 :: TEXT AS "related_type",
    $8 :: TEXT AS "type",
    NULL :: json AS "data",
    "public"."favorite_color"."color" AS "sort_color"
  FROM
    "public"."favorite_color"
    INNER JOIN "cte_peoples_data_0" ON "cte_peoples_data_0"."rel_favorite_color" = "public"."favorite_color"."id"
),
"cte_final_main_data" AS (
  SELECT
    DISTINCT ON ("cte_peoples_data_0"."main_id") "cte_peoples_data_0".*,
    "cte_peoples_data_0"."sort_first-name" AS "sort__first-name",
    "cte_rel_favorite_color_data_1"."sort_color" AS "sort_favorite_color_color"
  FROM
    "cte_peoples_data_0"
    LEFT JOIN "cte_rel_favorite_color_data_1" ON "cte_rel_favorite_color_data_1"."main_id" = "cte_peoples_data_0"."rel_favorite_color"
  ORDER BY
    "cte_peoples_data_0"."main_id"
) (
  SELECT
    "cte_final_main_data"."id",
    "cte_final_main_data"."type",
    "cte_final_main_data"."data",
    "cte_final_main_data"."related_type",
    "cte_final_main_data"."related_id"
  FROM
    "cte_final_main_data"
  ORDER BY
    "cte_final_main_data"."sort_favorite_color_color" DESC,
    "cte_final_main_data"."sort__first-name" ASC
);
